[Merged by Bors] - feat: a `grw` tactic for rewriting using inequalities

sebzim4500 opened the pull request:
Co-authored-by: @hrmacbeth, @digama0  

---
<!-- The text above the `---` will become the commit message when your
PR is merged. Please leave a blank newline before the `---`, otherwise
GitHub will format the text above it as a title.

To indicate co-authors, include lines at the bottom of the commit message
(that is, before the `---`) using the following format:

Co-authored-by: Author Name <author@email.com>

Any other comments you want to keep out of the PR commit should go
below the `---`, and placed outside this HTML comment, or else they
will be invisible to reviewers.

If this PR depends on other PRs, please list them below this comment,
using the following format:
- [ ] depends on: #abc [optional extra text]
- [ ] depends on: #xyz [optional extra text]
-->
- [x] depends on: #8796
```lean
example (x y z w : ℝ) (h₁ : x < z) (h₂ : w ≤ y + 4) (h₃ : Real.exp z < 5 * w) :
    Real.exp x < 5 * (y + 4) := by
  grw [h₁, ← h₂]
  exact h₃
```

[![Open in Gitpod](https://gitpod.io/button/open-in-gitpod.svg)](https://gitpod.io/from-referrer/)

semorrison commented on Mathlib/Tactic/GRW.lean:
```diff
@@ -0,0 +1,114 @@
+      let s ← withRef rule do
+        let symm := !rule[0].isNone
+        let term := rule[1]
+        -- let processId (id : Syntax) : TacticM Unit := do
```
```suggestion
        -- let processId (id : Syntax) : TacticM Unit := do
```
remove?

semorrison commented on Mathlib/Tactic/GRW/Lemmas.lean:
```diff
@@ -0,0 +1,43 @@
+lemma rewrite_mem {α : Type} {a : α} {X Y: Set α} (h₁ : a ∈ X) (h₂ : X ⊆ Y) : a ∈ Y := h₂ h₁
+
+@[grw]
+lemma rewrite_sub {α : Type} {X Y Z W: Set α} (h₁ : X ⊆ Y) (h₂ : Z ⊆ X) (h₃ : Y ⊆ W) :
```
```suggestion
lemma rewrite_sub {α : Type} {X Y Z W : Set α} (h₁ : X ⊆ Y) (h₂ : Z ⊆ X) (h₃ : Y ⊆ W) :
```
etc

semorrison commented on Mathlib/Tactic/GRW/Lemmas.lean:
```diff
@@ -0,0 +1,43 @@
+
+-/
+
+open Mathlib.Tactic.GRW
```
These are only for internal use.
```suggestion
namespace Mathlib.Tactic.GRW
```

semorrison commented on Mathlib/Tactic/GRW/Core.lean:
```diff
@@ -0,0 +1,197 @@
+  for lem in lemmas do
+    let lemResult : Option (Expr × Array MVarId)
+        ← withTraceNode `GRW (λ _ ↦ return m!"trying lemma {lem}") do
+      let s ← saveState
```
Does `commitIfNoEx` do what you want here?

semorrison commented on Mathlib/Tactic/GRW/Core.lean:
```diff
@@ -0,0 +1,197 @@
+
+  return rule
+
+/--
+Use the relation `rule` to rewrite `expr`
+-/
```
Could you add some explanation of the `isTarget` parameter?

semorrison commented on Mathlib/Tactic/GRW.lean:
```diff
@@ -0,0 +1,114 @@
+      ).run (⟨← getMainGoal, #[]⟩ : MVarId × Array MVarId)
+      try newGoal.withContext $ withReducible newGoal.applyRfl
+      catch _ => pure ⟨⟩
+      -- We can't use replaceMainGoal, since withTacticInfoContext prunes the solves goals so there
```
```suggestion
      -- We can't use replaceMainGoal, since withTacticInfoContext prunes the solved goals so there
```

semorrison commented on Mathlib/Tactic/GRW.lean:
```diff
@@ -0,0 +1,114 @@
+-/
+elab tok:"grw" rules:rwRuleSeq loc:(location)? : tactic => do
+  withLocation (expandOptLocation (Lean.mkOptionalNode loc))
+    (atLocal := λ fvar => withMainContext do
+      let ⟨_, newGoal, subgoals, _⟩ ← (withRWRulesSeqState tok rules fun rev syn ↦ do
+        let ⟨goal, subgoals, fvar⟩ ← get
+        goal.withContext do
+          let rulePrf ← elabTerm syn none
+          let ⟨newType, newHyp, newSubgoals⟩ ← goal.withContext
+              $ grwHypothesis (Expr.fvar fvar) rulePrf rev
+          let name ← fvar.getUserName
+          let ⟨newFvar, goal', _⟩ ← goal.assertAfter fvar name newType newHyp
+          let newGoal ← goal'.clear fvar
+          set (⟨newGoal, subgoals ++ newSubgoals, newFvar⟩ : MVarId × Array MVarId × FVarId)
+      ).run (⟨← getMainGoal, #[], fvar⟩ : MVarId × Array MVarId × FVarId)
+      let newGoals := subgoals ++ #[newGoal] ++ (← getGoals)
+      setGoals newGoals.toList
+      pruneSolvedGoals
+    )
```
I would recommend splitting out the `atLocal` and `atTarget` code into separate declarations (possibly at `MetaM` rather than `TacticM`). It will be more comprehensible, and down the track someone will probably want the MetaM plumbing available for use inside another tactic.

semorrison commented:
The PR description here says it is still a work in progress. Do you want to update that?

sebzim4500 commented:
> The PR description here says it is still a work in progress. Do you want to update that?

Yes. There are still features I want to add in the future but this PR is mergeable in my opinion.

digama0 commented:
Hi @sebzim4500 , today @hrmacbeth and I sat down and reviewed this code, resulting in some modifications and improvements in 329cbf51 that I have taken the liberty of pushing directly to the branch. Don't hesitate to speak up if you disagree or want clarification on the changes here. Summary of changes: 

* Use `IO.Ref` for persisting state through `withRwRulesSeq` instead of adding a `StateT` variant of the function
* Make `grw` transparently call `rw` when the hypothesis is an equality
  * thus `withRwRulesSeqState` becomes redundant and can be removed
* Inline/remove `grwHyp` into `grwAtLocal` and `grw` into `grwAtTarget`, rename `runGrw` to `grw`
* To avoid making too many assumptions about the inputs, `grw` core does not change any mvars, instead it returns a proof whose type is an implication in the specified direction.
* Interchange `withRwRulesSeq` and `withLocation`; this means `grw [a, b] at h1 h2` is equivalent to `grw [a] at h1 h2; grw [b] at h1 h2` instead of `grw [a, b] at h1; grw [a, b] at h2`, which matches how `rw` works and also makes the intermediate states between commas more sensible. (Previously using `grw [a, b] at *` would generate a large number of goals in the intermediate states because the same syntax is being tagged many times.)
* Use tactic state for persisting information rather than an `IO.Ref` or `StateT`. Besides being simpler to implement, this is also consistent with the `rw` tactic.
  * Downside: we do not track the fvar of the hypothesis being rewritten, so in theory a different hypothesis with the same name can jump in halfway through a rewrite chain (although examples that do this are extremely contrived)
  * Downside: Side goals come out in reverse order, i.e. we have `main goal, side goals for b, side goals for a` when doing `grw [a, b]`
  * Both of these are consistent with the `rw` tactic though, which we considered more important.
* Pass a template to `gcongr`, to improve the speed (and also catch some bugs).
* `dischargeSideGoal` is the same discharger used in `gcongr` itself, this is made the default in #8796
* `dischargeMainGoal` is replaced by `gcongrForward` with the specified `rule`
  * `gcongrForward` overlaps significantly with the `grw_weaken` attribute; the two have been merged and `grw_weaken` removed.
* The `progress` return value of `gcongr` is ignored - we don't really care if `gcongr` does nothing because it is being used on subgoals
* remove unnecessary partial and private from definitions
* add more docs
* Added some more tests
* reduced imports of test file

@hrmacbeth will push some more test files shortly but otherwise we support merging it. There are still some TODOs remaining, but they can be addressed in a follow-up PR:

* The `grw` lemmas are stored in a big bag and all applied to the goal. This should be replaced with some discrimination tree indexing or a lookup on the head symbol like in `gcongr`, this implementation does not scale and will likely become a problem if this tactic starts getting used widely (which we expect to happen, it's a very useful tactic).
* The attribute is not checked at application time, it is just a `LabelAttr`. We would like it to enforce some structure, both to make it easier for the user to apply the tag correctly, and also to improve the performance and precision of the tactic itself; currently it has some hacks to guess the structure of the lemma but these can be easily fooled.

hrmacbeth commented on Mathlib/Tactic/GRW.lean:
```diff
@@ -0,0 +1,74 @@
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Sebastian Zimmer, Mario Carneiro, Heather Macbeth
+-/
+import Mathlib.Lean.Meta
+import Mathlib.Lean.Expr
+import Mathlib.Lean.Elab.Tactic.Basic
+import Lean.Parser.Tactic
+import Lean.Meta.Tactic
+import Std.Tactic.LabelAttr
+import Mathlib.Tactic.GRW.Core
```
```suggestion
```

eric-wieser commented on Mathlib/Tactic/GRW/Core.lean:
```diff
@@ -0,0 +1,179 @@
+
+open GCongr
+
+/-- Given `relationType := R a b`, returns `(a, b)` -/
```
Should this be counting explicit arguments? I'm considering a relation which has some trailing typeclass arguments (as an extreme case, `MonoidHom`!)

eric-wieser commented on Mathlib/Tactic/GRW/Core.lean:
```diff
@@ -0,0 +1,179 @@
+`rule` for discharging the main subgoals. -/
+def useRule (template rule : Expr) (goal : MVarId) : MetaM (Array MVarId) := do
+  let template ← instantiateMVars template
+  try
+    if template.hasExprMVar then failure
+    goal.applyRfl
+    trace[GRW] "used reflexivity"
+    return #[]
+  catch _ => pure ()
```
This is going to catch all sorts of unwanted things like timeouts, isn't it?

eric-wieser replied:
It looks like there's no good answer here (though maybe @digama0 knows of one).

digama0 replied:
this is not really something that should be solved here, it's endemic to the whole tactic infrastructure. Maybe we could change the `tryCatch` implementation for `TacticM` to not catch timeouts and other abort-level commands?

eric-wieser replied:
I think that must be a recent change: thanks for digging!

eric-wieser commented on Mathlib/Tactic/GRW/Lemmas.lean:
```diff
@@ -0,0 +1,40 @@
+    X' ⊂ Y' := lt_of_le_of_lt hX (lt_of_lt_of_le H hY)
+
+@[grw] lemma rewrite_bddAbove {α : Type*} [Preorder α] {s s' : Set α}
+    (H : BddAbove s) (hs : s' ⊆ s) : BddAbove s' := H.imp fun _ hc _ hx => hc (hs hx)
```
I would assume this can be proven with `BddAbove.mono`? Or is that not imported?

Vierkantor commented on Mathlib/Tactic/GRW/Core.lean:
```diff
@@ -0,0 +1,179 @@
+      if let some (lemExpr, metas) := lemResult then
+        let metas ← metas.filterM fun x => not <$> x.mvarId!.isAssigned
+        let args := template.getAppArgs
+        -- HACK: we are assuming the side goals of the grw lemma come in the same order
+        -- as they appear in the function, and all the fixed args (e.g. type, instances)
+        -- come before the ones to rewrite
```
I'm not completely sure what you mean by "as they appear in the function" here. What are the effects of violating this assumption? Can it be checked when tagging a lemma as `@[grw]`?

digama0 replied:
Yes, this was planned to be checked in the attribute. The [HM-grw-attribute](https://github.com/leanprover-community/mathlib4/commit/ef82c2de29ac2f582687c38f809f199ed3373698) branch implements that.

digama0 commented on Mathlib/Tactic/GRW/Core.lean:
```diff
@@ -0,0 +1,179 @@
+  let lemmas ← labelled `grw
+  let (lhsExpr, rhsExpr) := if isTarget then (newExpr, oldExpr) else (oldExpr, newExpr)
+
+  -- TODO surely this can be faster
+  for lem in lemmas do
```
@hrmacbeth already has a fixed version of the TODO at https://github.com/leanprover-community/mathlib4/commit/ef82c2de29ac2f582687c38f809f199ed3373698 , although it has not been merged with this PR yet.

fpvandoorn commented on Mathlib/Tactic/GRW.lean:
```diff
@@ -0,0 +1,67 @@
+If applied to the target then the tactic will attempt to close the goal with `rfl` after doing the
+rewriting.
+-/
+elab tok:"grw" rules:rwRuleSeq loc:(location)? : tactic => do
```
I believe `Data.Set.Basic` is "too heavy" to import in `Mathlib.Tactic.Common`?

sebzim4500 commented:
I'm coming back to this project now.

I've rebased this branch, made the requested changes and fixed the tests.

@hrmacbeth I see mentions of a separate branch that I don't think was ever merged into this one? Does that still exist? I realize it was a long time ago.

fpvandoorn commented:
I would love to see this in Mathlib. Github has a feature that you can see all the branches a commit is part of, so maybe another mentioned branch is one of the 4 listed at the top of this page? https://github.com/leanprover-community/mathlib4/commit/fb88bd8b5dabdfd415c62dc6e1683f5db6668ae3

fpvandoorn reviewed (commented):
I didn't look at the implementation yet, but here are some comments.

sebzim4500 commented on MathlibTest/GRW.lean:
I think this is only the case when you use it in a way where it doesn't end up in the final proof term. I don't think it has anything to do with the `grw` tactic, you can get the same behavior with e.g. the `rw` tactic.

It really only happens in tests because normally there'd be no reason to do rewrites in expressions which you throw away.

fpvandoorn replied:
Oh, I see. If that is the case, feel free to turn off the linter in the test file, with this remark.

sebzim4500 replied:
How do you disable a lint for an entire file and not just one definition?

fpvandoorn replied:
`set_option.linter.unused_variable false` (or something similar) near the top.

fpvandoorn commented on Mathlib/Tactic/GRW/Lemmas.lean:
```diff
@@ -0,0 +1,42 @@
+    (H : a < b) (ha : a' ≤ a) (hb : b ≤ b') : a' < b' := lt_of_le_of_lt ha (lt_of_lt_of_le H hb)
+
+@[grw] lemma rewrite_mem {α : Type*} {a : α} {X X' : Set α}
+    (H : a ∈ X) (hX : X ⊆ X') (_ : a = a) : a ∈ X' := hX H
```
It's possible that I didn't look closely enough and missed the error... In that case, use the `nolint` attribute.

sebzim4500 commented on Mathlib/Tactic/GRW/Lemmas.lean:
I tried to make it more flexible but it ended up being much slower.

Given I don't expect there to that many more of these lemmas than the ones already in this PR (and 99% of usage will probably just use the ones on < and <=) I don't think it's that big a deal.

JovanGerb commented:
I have a suggestion for refactoring the code:
- We can generalize `gcongr` to also be able to deal with the implication relation (`· → ·`). We need to teach this to the function that interprets an expression as an application of a relation. And it may require a separate `@[gcongr_imp]` for tagging `gcongr` lemmas about the `· → ·` relation.
- Then `@[gcongr_imp]` is a replacement for `@[grw]`.
- This should make the implementation a lot shorter, because we can call `gcongr` directly on the goal `newExpr → oldExpr` or `oldExpr → newExpr`.

And as a bonus, we can gain the ability to rewrite using implications.

JLimperg commented:
A student of mine, @FWuermse, has implemented [another `grw` tactic](https://github.com/FWuermse/grw). This one copies the Rocq approach to generalised rewriting, described in [this paper](https://jfr.unibo.it/article/view/1574), including various optimisations from the Rocq codebase that make the algorithm practical. It seems to me like the Rocq algorithm is much more general than the one proposed here.

Florian's tactic is maybe 80% done, with some minor functionality missing and probably some cleanup required. It's not clear how much time Florian will be able to put into it, and I'm very busy these days, so I don't want to promise anything and I don't want to scupper this PR. But it might be worth thinking about what should happen if Florian's tactic ever becomes ready for Mathlib.

fpvandoorn commented:
What are examples of the Rocq-version that this one cannot do?

leanprover-community-bot-assistant commented:
This pull request has conflicts, please merge `master` and resolve them.
